GH-25986: [C++] Enable external material and rotation for encryption keys

adamreeve opened the pull request:
This PR is a replacement for #10491 which appears to be abandoned. I've fixed the issues pointed out in review comments as well as a few more things I noticed. I've also made one significant change to the API; rather than `CryptoFactory::RotateMasterKeys` operating on a whole directory, it rotates keys for a single file, as this is much more flexible and allows users to decide what files to rotate keys for, and there didn't appear to be a good reason for this to work at the whole directory level. This does mean the API diverges slightly from the parquet-mr API though.

### Rationale for this change

Use of external key material allows rotating master encryption keys without having to rewrite Parquet file data. See https://docs.google.com/document/d/1bEu903840yb95k9q2X-BlsYKuXoygE4VnMDl9xz_zhk/edit?usp=sharing for more details.

### What changes are included in this PR?

Adds support for writing and reading external key material for Parquet files from C++, as well as a new `CryptoFactory::RotateMasterKeys` function that allows re-encrypting key encryption keys or data encryption keys with latest versions of master keys.

### Are these changes tested?

Yes, unit tests are included. I've added an additional test that reads a file generated with parquet-mr (used by Spark) from the parquet-testing repository. This requires merging the PR at https://github.com/apache/parquet-testing/pull/36 and updating the parquet-testing submodule before the new test will pass.

### Are there any user-facing changes?

Yes, the existing `internal_key_material` option in `parquet::encryption::EncryptionConfiguration` will now work and use external key material. This requires using two new parameters (`file_path` and `file_system`) in `CryptoFactory::GetFileEncryptionProperties` and `CryptoFactory::GetFileDecryptionProperties`, which are needed so that we know where to write/read the external key material. Note that this means external key material won't work from Python until the new parameters are exposed in Python too.

This changes the `CryptoFactory` ABI but the API is still source compatible. Is this okay or should new overloads be added that take the new parameters?

The `CryptoFactory::RotateMasterKeys` function is also a new public facing API.
* Closes: #25986

wjones127 commented on cpp/src/parquet/encryption/key_management_test.cc:
```diff
@@ -122,45 +132,87 @@ class TestEncryptionKeyManagement : public ::testing::Test {
+      encryptor_.EncryptFile(file, file_encryption_properties);
+    } else {
+      auto file_system = std::make_shared<::arrow::fs::LocalFileSystem>();
+      std::string file_path = temp_dir_->path().ToString() + file_name;
```
This also the same thing as `file` above?

wjones127 reviewed (changes requested):
This is looking pretty good. I have a few minor comments and then I think we are waiting on https://github.com/apache/parquet-testing/pull/36

adamreeve commented on cpp/src/parquet/encryption/file_key_unwrapper.h:
```diff
@@ -44,14 +43,21 @@ class PARQUET_EXPORT FileKeyUnwrapper : public DecryptionKeyRetriever {
   /// KmsClient if it's not in the cache yet. cache_entry_lifetime_seconds is life time of
   /// KmsClient in the cache.
+  /// If the file uses external key material then either the Parquet file path and file
+  /// system must be specified, or a key material store provided.
   FileKeyUnwrapper(KeyToolkit* key_toolkit,
                    const KmsConnectionConfig& kms_connection_config,
-                   double cache_lifetime_seconds);
+                   double cache_lifetime_seconds, const std::string& file_path,
+                   const std::shared_ptr<::arrow::fs::FileSystem>& file_system = NULLPTR,
+                   std::shared_ptr<FileKeyMaterialStore> key_material_store = NULLPTR);
```
It might make sense for `CryptoFactory::GetFileEncryptionProperties` and `CryptoFactory::GetFileDecryptionProperties` to have separate overloads as they seem closer to the public API that most people will interact with? I'm not so sure about this constructor, it's in the public encryption namespace but does seem closer to the internals, and is only called from `CryptoFactory::GetFileDecryptionProperties` within the arrow source, where we pass all parameters.

adamreeve commented:
Thanks for the feedback @wjones127, I've addressed all of these points except for the one about using overloads as I think it might be more important to have overloads on the `CryptoFactory` methods instead. I'm not sure how important ABI stability is though, should anything not in an internal namespace use overloads when adding parameters to avoid breaking ABI compatibility?

I've also updated the parquet-testing submodule now that https://github.com/apache/parquet-testing/pull/36 is merged, so all tests should pass :crossed_fingers:

adamreeve commented:
> I think overloads seem appropriate for the methods that take _either_ `KeyMaterialStore` or file path + `FileSystem`. It seems weird to have one function that expects only one of either parameters to be passed.

Right, sorry I totally missed the point of your comment. I've split this constructor into two overloads now.

wjones127 reviewed (approved):
This looks good to me now. If there are no other comments, I'll merge this Monday.

ggershinsky commented:
@adamreeve Thank you very much for this PR. Do you know if there are plans to make this work from Python? (at least the read part)

adamreeve commented:
Hi @ggershinsky, no, I don't have any plans to make this work from Python, and I'm not aware of any. It's a while since I worked on this, but I don't think there would be too much work required, mainly just adding the file path and filesystem arguments to `CryptoFactory.file_decryption_properties` and  `CryptoFactory.file_encryption_properties`.

ggershinsky commented:
Sounds good, we'll have a look.
